src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -296,5 +284,160 @@ internal SafeSslHandle(IntPtr validSslPointer, bool ownsHandle) : base(IntPtr.Ze
         {
             handle = validSslPointer;
         }
+
+        internal class ReadBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private int _bytesAvailable;
+            private byte[] _byteArray;
+            private int _offset;
+
+            internal ReadBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.Crypto.ManagedSslBio.BioSetGCHandle(_bioHandle, _handle);
+                Interop.Crypto.BioSetShoudRetryReadFlag(bioHandle);
+            }
+
+            public void SetData(byte[] buffer, int offset, int length)
+            {
+                Debug.Assert(_bytesAvailable == 0);
+
+                _byteArray = buffer;
+                _offset = offset;
+                _bytesAvailable = length;
+            }
+
+            public int Read(Span<byte> output)
+            {
+                var bytesToCopy = Math.Min(output.Length, _bytesAvailable);
+                if (bytesToCopy == 0)
+                {
+                    return -1;
+                }
+
+                var span = new Span<byte>(_byteArray, _offset, bytesToCopy);
+                span.CopyTo(output);
+                _offset += bytesToCopy;
+                _bytesAvailable -= bytesToCopy;
+                return bytesToCopy;
+            }
+
+            //Bio is already released by the ssl object
+            private void Dispose(bool isDisposing)
+            {
+                if (_handle.IsAllocated)
+                {
+                    _handle.Free();
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+
+            ~ReadBioBuffer()
+            {
+                Dispose(false);
+            }
+        }
+
+        internal class WriteBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private byte[] _byteArray;
+            private int _bytesWritten;
+            private bool _isHandshake;
+
+            internal WriteBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.Crypto.ManagedSslBio.BioSetGCHandle(_bioHandle, _handle);
+            }
+
+            public void SetData(byte[] buffer, bool isHandshake)
+            {
+                _byteArray = buffer;
+                _bytesWritten = 0;
+                _isHandshake = isHandshake;
+            }
+
+            public int TakeBytes(out byte[] output)
+            {
+                output = _byteArray;
+                return TakeBytes();
+            }
+
+            public int TakeBytes()
```

Comment:
It looks like it's just used to null out the array after a check that no work remains.  The function that calls it already knows the return value (as its `retVal` local), so this doesn't seem to need the somewhat confusing return value.

If we're nulling out the array just for GC purposes, then perhaps `ClearData` is a better name for the public version?